File: frontend/src/mahjong-puzzle/tiles.ts
import { Pai, comparePai } from '@/mahjong-calc/types';

/** 谜题用牌序（不含赤宝牌） */
export const PUZZLE_TILE_ORDER = [
  '1m', '2m', '3m', '4m', '5m', '6m', '7m', '8m', '9m',
  '1p', '2p', '3p', '4p', '5p', '6p', '7p', '8p', '9p',
  '1s', '2s', '3s', '4s', '5s', '6s', '7s', '8s', '9s',
  '1z', '2z', '3z', '4z', '5z', '6z', '7z',
] as const;

export const PUZZLE_TILE_ROWS = [
  ['1m', '2m', '3m', '4m', '5m', '6m', '7m', '8m', '9m'],
  ['1p', '2p', '3p', '4p', '5p', '6p', '7p', '8p', '9p'],
  ['1s', '2s', '3s', '4s', '5s', '6s', '7s', '8s', '9s'],
  ['1z', '2z', '3z', '4z', '5z', '6z', '7z'],
] as const;

export type PuzzleTile = (typeof PUZZLE_TILE_ORDER)[number];

export function tileToIndex(tile: string): number {
  const m = tile.match(/^(\d)([mpsz])$/);
  if (!m) return -1;
  const num = parseInt(m[1]!, 10);
  const suit = m[2]!;
  if (suit === 'm') return num - 1;
  if (suit === 'p') return 9 + num - 1;
  if (suit === 's') return 18 + num - 1;
  if (suit === 'z') return 27 + num - 1;
  return -1;
}

export function tileToPai(tile: string): Pai {
  const m = tile.match(/^(\d)([mpsz])$/);
  if (!m) throw new Error(`invalid tile: ${tile}`);
  return new Pai(m[2] as Pai['type'], parseInt(m[1]!, 10));
}

export function paiToTile(pai: Pai): string {
  return `${pai.num}${pai.type}`;
}

export function sortTilesCanonical(tiles: string[]): string[] {
  return [...tiles].sort((a, b) => comparePai(tileToPai(a), tileToPai(b)));
}

/** 标准答案：13 张手牌按默认顺序 + 第 14 格为摸牌 */
export function buildCanonicalAnswer(hand13: string[], draw: string): string[] {
  return [...sortTilesCanonical(hand13), draw];
}

export function countTiles(tiles: string[]): Record<string, number> {
  const c: Record<string, number> = {};
  for (const t of tiles) {
    if (t) c[t] = (c[t] ?? 0) + 1;
  }
  return c;
}

export function tilesToC34(tiles: string[]): number[] {
  const c34 = new Array(34).fill(0);
  for (const t of tiles) {
    const i = tileToIndex(t);
    if (i >= 0) c34[i]++;
  }
  return c34;
}

File: backend/quemi/tiles.go
package quemi

import (
	"fmt"
	"regexp"
	"sort"
)

var tilePattern = regexp.MustCompile(`^(\d)([mpsz])$`)

// PuzzleTileOrder is the canonical tile sort order (no red fives).
var PuzzleTileOrder = []string{
	"1m", "2m", "3m", "4m", "5m", "6m", "7m", "8m", "9m",
	"1p", "2p", "3p", "4p", "5p", "6p", "7p", "8p", "9p",
	"1s", "2s", "3s", "4s", "5s", "6s", "7s", "8s", "9s",
	"1z", "2z", "3z", "4z", "5z", "6z", "7z",
}

// TileToIndex maps a tile string to 0–33 index, or -1 if invalid.
func TileToIndex(tile string) int {
	m := tilePattern.FindStringSubmatch(tile)
	if m == nil {
		return -1
	}
	num := int(m[1][0] - '0')
	switch m[2] {
	case "m":
		return num - 1
	case "p":
		return 9 + num - 1
	case "s":
		return 18 + num - 1
	case "z":
		return 27 + num - 1
	default:
		return -1
	}
}

// TileToPai parses a tile string into Pai.
func TileToPai(tile string) (Pai, error) {
	m := tilePattern.FindStringSubmatch(tile)
	if m == nil {
		return Pai{}, fmt.Errorf("invalid tile: %s", tile)
	}
	return Pai{Type: PaiType(m[2]), Num: int(m[1][0] - '0')}, nil
}

// MustTileToPai parses a tile or panics (for internal use with known-good tiles).
func MustTileToPai(tile string) Pai {
	p, err := TileToPai(tile)
	if err != nil {
		panic(err)
	}
	return p
}

// PaiToTile formats Pai as tile string.
func PaiToTile(p Pai) string {
	return fmt.Sprintf("%d%s", p.Num, p.Type)
}

// SortTilesCanonical returns tiles sorted by comparePai order.
func SortTilesCanonical(tiles []string) []string {
	out := append([]string(nil), tiles...)
	sort.Slice(out, func(i, j int) bool {
		return ComparePai(MustTileToPai(out[i]), MustTileToPai(out[j])) < 0
	})
	return out
}

// BuildCanonicalAnswer returns 13 sorted hand tiles + draw.
func BuildCanonicalAnswer(hand13 []string, draw string) []string {
	sorted := SortTilesCanonical(hand13)
	out := make([]string, 0, 14)
	out = append(out, sorted...)
	out = append(out, draw)
	return out
}

// CountTiles counts occurrences per tile string.
func CountTiles(tiles []string) map[string]int {
	c := make(map[string]int)
	for _, t := range tiles {
		if t == "" {
			continue
		}
		c[t]++
	}
	return c
}

// TilesToC34 converts tiles to a 34-element count array.
func TilesToC34(tiles []string) [34]int {
	var c34 [34]int
	for _, t := range tiles {
		i := TileToIndex(t)
		if i >= 0 {
			c34[i]++
		}
	}
	return c34
}
